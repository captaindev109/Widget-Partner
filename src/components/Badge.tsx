import { FC } from 'react';
import styled from 'styled-components';
import Card from '@/generics/Card';

export interface BadgeProps {
  children: string;
}

const StyledBadge = styled(Card)<BadgeProps>`
  background-color: ${(props) => props.theme.color.purple300};
  font-family: ${(props) => props.theme.fonts.FAVORIT};
  border-radius: 4px;
  padding: 5px 8px 3px;
  line-height: 16px;
  font-size: 12px;
  color: ${(props) => props.theme.color.gray900};
`;

const Badge: FC<BadgeProps> = ({ children }) => {
  return <StyledBadge data-testid="badge">{children.toUpperCase()}</StyledBadge>;
};

const defaultProps: Partial<BadgeProps> = {};

Badge.defaultProps = defaultProps;

export default Badge;
