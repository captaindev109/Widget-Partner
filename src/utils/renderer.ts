import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
