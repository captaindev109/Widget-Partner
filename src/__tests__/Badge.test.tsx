import { render, RenderResult } from '@testing-library/react';
import queries from '@testing-library/dom/types/queries';
import renderer from 'react-test-renderer';
import Badge from '../components/Badge';

import { FC } from 'react';
import { ThemeProvider } from 'styled-components';
import defaultTheme from '@/theme/theme';

const child_text = 'child';

const BadgeComponent: FC = () => (
  <>
    <ThemeProvider theme={defaultTheme}>
      <Badge>{child_text}</Badge>
    </ThemeProvider>
  </>
);

const STYLE_PROPS: Record<string, string> = {
  'border-radius': '4px',
  padding: '5px 8px 3px',
  'line-height': '16px',
  'font-size': '12px',
};

describe('Test cases of the Badge component', () => {
  let renderedComponent: RenderResult<typeof queries, HTMLElement>;

  beforeEach(() => {
    renderedComponent = render(<BadgeComponent />);
  });

  test('to see if the badge component render without crashing', () => {
    render(<BadgeComponent />);
  });

  test('to see if the Badge text render uppercase', () => {
    const { getByTestId } = renderedComponent;
    const badge = getByTestId('badge');

    expect(badge).toHaveTextContent(child_text.toUpperCase());
  });

  test('to see if the Badge styling is correct', () => {
    const { getByTestId } = renderedComponent;
    const badge = getByTestId('badge');

    // Mapping through the object to check if the style props are applied
    Object.entries(STYLE_PROPS).forEach(([key, value]) => {
      expect(badge).toHaveStyle(`${key}: ${value}`);
    });
  });

  it('should match the sapshot', () => {
    const snap = renderer.create(<BadgeComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
