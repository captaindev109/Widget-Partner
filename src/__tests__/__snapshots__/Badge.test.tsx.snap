// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cases of the Badge component should match the sapshot 1`] = `
<div
  className="Card-sc-fb64gk-0 Badge__StyledBadge-sc-12y41wr-0 leCJCh hwjXFa"
  data-testid="badge"
>
  CHILD
</div>
`;
